### DISCUSSION TITLE
Is it possible to persist using Svelte right now?

### DISCUSSION THREAD

**Discussion by frederikhors** [TanStack/query#5378] [Q&A] [answered]
Svelte lover here: is it possible [to persist](https://tanstack.com/query/latest/docs/react/plugins/persistQueryClient) with Svelte too?

Or is just for React for now?

If not this is a blocking for us unfortunately.

**Comment by TkDodo** [comment]
The react adapter is pretty slim, you can take a look here: 

https://github.com/TanStack/query/blob/f57c8dc15ec7cc20c674892eacd7596a4df39085/packages/react-query-persist-client/src/PersistQueryClientProvider.tsx#L21-L53

I think you can do that in svelte in user-land and then maybe contribute it back as a `svelte-query-persist-client` ? @lachlancollins FYI

**Reply by frederikhors** [reply]
I don't think I'm the perfect guy for such a job, unfortunately.

**Comment by ghost** [comment, marked as answer]
@frederikhors, i was able to persist and load client with this when disabling ssr ( might also work with ssr). Let me know if this worked? BTW I'm not using the query-persist-core because it rewrites every time the query cache changes, and I think frequently JSON.stringify-ing and writing the local storage might slows down my app ( I maybe wrong here ), so what i did here was saved the query state on unload. This only works when the function is non asynchronous and writing to local storage is synchronous.  

```svelte
<script>
  import { setContext, onMount } from "svelte";
  import { hydrate, dehydrate } from "@tanstack/svelte-query";
  import { SvelteQueryDevtools } from "@tanstack/svelte-query-devtools";

  export let data;

  const queryClient = data.queryClient;
  setContext("$$_queryClient", queryClient);

  const hydrateClient = () => {
    try {
      const storeValue = localStorage.getItem("QUERY_CLIENT");
      if (!storeValue) return;

      const persistedValue = JSON.parse(storeValue);

      if (persistedValue?.timestamp) {
        const MAX_AGE = 1000 * 60 * 60 * 24;
        const expired = Date.now() - persistedValue.timestamp > MAX_AGE;
        if (!expired) hydrate(queryClient, persistedValue.clientState);
      } else {
        localStorage.removeItem("QUERY_CLIENT");
      }
    } catch {
      // remove persisted client on any error
      localStorage.removeItem("QUERY_CLIENT");
    }
  };
  const saveClient = () => {
    localStorage.setItem(
      "QUERY_CLIENT",
      JSON.stringify({
        timestamp: Date.now(),
        clientState: dehydrate(queryClient, {})
      })
    );
  };

  const unload = () => {
    saveClient();
  };

  onMount(() => {
    hydrateClient();
    queryClient.mount();
    return () => {
      queryClient.unmount();
    };
  });
</script>

<slot />
<SvelteQueryDevtools buttonPosition="bottom-right" />
<svelte:window on:beforeunload={unload} />
```

**Reply by ghost** [reply]
and about IsRestoringProvider, i didn't find any usage for it. Although if you need it, you can create a svelte store ```const restoring = writable(false)``` and set it to true within try block of hydrateClient. Then add a finally block to set it back to false
```js
const hydrateClient = () => {
    try {
        restoring.set(true);
       // rest of the above
    } catch {
      // rest of the catch block
    } finally {
      restoring.set(false)
   }
}
```

I dont know if these are good ways for persisting query, I'm not that expert on javascript, but the above is working for me and hope this might help you :smiley:

**Reply by frederikhors** [reply]
Congratulations on the idea and goodwill.

> I'm not using the query-persist-core because it rewrites every time the query cache changes, and I think frequently JSON.stringify-ing and writing the local storage might slows down my app ( I maybe wrong here ), so what i did here was saved the query state on unload. This only works when the function is non asynchronous and writing to local storage is synchronous.

This is the issue here. I would like to use `IndexedDB` because it is ASYNC and not sync (like `localstorage` is). And because the `localstorage` space is limited (~ 5 MB only).

[urql is using IndexedDB by default for this very reason](https://formidable.com/open-source/urql/docs/graphcache/offline/#custom-storages).

**Reply by ghost** [reply]
you can use the query core persist. Are you using sveltekit or only svelte?

**Reply by ghost** [reply]
for sveltekit, the code would be something similar to:
```svelte
<script>
  import { get, set, del } from "idb-keyval";
  import { setContext, onMount } from "svelte";
  import { SvelteQueryDevtools } from "@tanstack/svelte-query-devtools";
  import { persistQueryClient } from "@tanstack/query-persist-client-core";
  import type { PersistedClient, Persister } from "@tanstack/query-persist-client-core";

  export let data;

  let restoring = false;
  const queryClient = data.queryClient;
  setContext("$$_queryClient", queryClient);

  const persister: Persister = {
    async persistClient(client: PersistedClient) {
      set("QUERY_PERSIST", client);
    },
    async restoreClient() {
      return await get<PersistedClient>("QUERY_PERSIST");
    },
    async removeClient() {
      await del("QUERY_PERSIST");
    }
  }

  onMount(() => {
    restoring = true;
    const [unsubscribe, promise] = persistQueryClient({ queryClient, persister });

    promise()
    .then(() => {
      queryClient.mount();
      // run paused mutations if you want
    })
    .finally(() => {
      restoring = false;
    });

    return () => {
      unsubscribe();
      queryClient.unmount();
    };
  });
</script>

<slot />
<SvelteQueryDevtools buttonPosition="bottom-right" />
```

**Reply by frederikhors** [reply]
@pukarsharma112 sorry for the brutal delay. Only now I've had the time to read your magnificent answer that I've tried and IT WORKS!

I'm very happy. Thank you very much.

Just a doubt:

- `promise()` is not callable, right?

- why are you using `queryClient.mount();`?

- are you using this code in `routes/+layout.ts`?

- why the `setContext("$$_queryClient", queryClient);`?

- why are you not using `QueryClientProvider` (from `@tanstack/svelte-query`)? Like:

    ```svelte
      <QueryClientProvider client={queryClient}>
        <slot />
      </QueryClientProvider>
    ```

**Reply by frederikhors** [reply]
@pukarsharma112 do you have any idea about https://github.com/TanStack/query/issues/5854?

**Reply by sharmapukar217** [reply]
Oh sorry, i got lost into my work.
a) promise is callable as far as i know.
b) I'm using queryClient.mount() because i wanted to load from cache first then handle rest like saved mutations so that I'd get fresh data as well,
c) Yes I'm using +layout.ts to create a new instance of queryClient. I'm using trpc so +layout.ts must be used for ssr stuffs
d & e) using QueryClientProvider won't allow me (b). 
Please correct me if i did something wrong up here 🙂

**Reply by sharmapukar217** [reply]
If promise was not callable, I'd have got error there. But the data were rehydrated from localstorage so i guess it's a function

**Reply by sharmapukar217** [reply]
Oh i forgot to mention, you can use `persistQueryClient` and provide a filter(I forgot exactly where, but if you want i could try it once for you) so that instead of rewriting every state of queryClient, it only saves successful and(or) error state, skipping loading state. This way, you can use idb for storage but with lesser rewrites

**Reply by frederikhors** [reply]
This is what I got if I try:

![image](https://github.com/TanStack/query/assets/41120635/a66c2196-e5b0-41ae-aeee-82ea5bac6429)

**Reply by frederikhors** [reply]
> you can use `persistQueryClient` and provide a filter

Can you please show how? I cannot understand how to do this.

**Reply by sharmapukar217** [reply]
Please give me few days. I'm away from keyboard since weeks.  I'd provide you details when I'm go back to my pc 🙂

**Reply by sharmapukar217** [reply]
I guess its promise.then(() => {}), i got confused maybe. You can check the source code for react wrapper.

**Comment by sharmapukar217** [comment]
![Screenshot_2023-08-12-21-46-41-095_com android chrome](https://github.com/TanStack/query/assets/82799213/01a3532b-f243-46b1-a3c2-863fe6f0e5bb)

Thats promise.then(() => {})
Sorry, my bad 😅
